### PR TITLE
Add multilayer LSTM for decoder + dropout on encoder and decoder

### PR DESCRIPTION
**Pull request by Ranapop**

### CODE REVIEW

File: examples/cfq/models.py [hunk 1]
```diff
@@ -34,12 +36,19 @@ def apply(self,
             inputs: jnp.array,
             lengths: jnp.array,
             shared_embedding: nn.Module,
-            hidden_size: int = LSTM_HIDDEN_SIZE):
+            train: bool,
+            hidden_size: int,
+            num_layers: int,
+            horizontal_dropout_rate: float,
+            vertical_dropout_rate: float):
```
Comment (marcvanzee) [inline examples/cfq/models.py]: Nice! I like that you renamed these, I think these names are clearer than the ones in recurrent.py.

File: examples/cfq/models.py [hunk 2]
```diff
@@ -91,18 +100,99 @@ def apply(self,
     return attention  # Shape: <float32>[batch_size, seq_len]
 
 
+class MultilayerLSTM(nn.Module):
+  "LSTM cell with multiple layers"
+
+  def apply(self,
+        num_layers,
+        horizontal_dropout_masks,
+        vertical_dropout_masks,
```
Comment (marcvanzee) [inline examples/cfq/models.py]: Please add a type

File: examples/cfq/models.py [hunk 3]
```diff
@@ -91,18 +100,99 @@ def apply(self,
...
+        layer (used only between layers, not on final output). The dropout is
+        the one used here https://arxiv.org/abs/1512.05287 except it's not
+        applied on the first layer inputs and final outputs to be consistent
+        with the implementation in `recurrent.py`
+        shape [num_layers-1, batch_size, hidden_size]
+      input: input given to the first LSTM layer [batch_size, input_size]
+      previous_states: list of (c,h) for each layer
+        shape [num_layers, batch_size, 2*hidden_size]
+    """
+    cell_input = input
```
Comment (marcvanzee) [inline examples/cfq/models.py]: This seem unnecessary. You can just use `input` everywhere instead of `cell_input`.

File: examples/cfq/models.py [hunk 4]
```diff
@@ -91,18 +100,99 @@ def apply(self,
...
+    states = []
+    final_output = None
+    for layer_idx in range(num_layers):
+      lstm_name = 'lstm_layer'+str(layer_idx)
+      cell = nn.LSTMCell.shared(name=lstm_name)
```
Comment (marcvanzee) [inline examples/cfq/models.py]: I don't think you should make this shared, since the cell is only used once and each time this is called the `lstm_name` is different.

File: examples/cfq/models.py [hunk 5]
```diff
@@ -91,18 +100,99 @@ def apply(self,
...
+      c,h = previous_states[layer_idx]
```
Comment (marcvanzee) [inline examples/cfq/models.py]:
```suggestion
      c, h = previous_states[layer_idx]
```

File: examples/cfq/models.py [hunk 6]
```diff
@@ -91,18 +100,99 @@ def apply(self,
...
+      # apply dropout to h
+      if horizontal_dropout_masks[layer_idx] is not None:
+        h = h * horizontal_dropout_masks[layer_idx]
+      # apply dropout to hidden state from lower layer
+      if layer_idx!=0 and vertical_dropout_masks[layer_idx-1] is not None:
```
Comment (marcvanzee) [inline examples/cfq/models.py]:
```suggestion
      # Apply dropout to the hidden state from the lower layer.
      if layer_idx != 0 and vertical_dropout_masks[layer_idx-1] is not None:
```

File: examples/cfq/models.py [hunk 7]
```diff
@@ -91,18 +100,99 @@ def apply(self,
...
+        cell_input = cell_input * vertical_dropout_masks[layer_idx-1]
+      state, output = cell((c,h),cell_input)
```
Comment (marcvanzee) [inline examples/cfq/models.py]:
```suggestion
      state, output = cell((c, h), cell_input)
```

File: examples/cfq/models.py [hunk 8]
```diff
@@ -91,18 +100,99 @@ def apply(self,
...
+
+
 class Decoder(nn.Module):
   """LSTM decoder."""
 
+  def create_dropout_masks(self,
+                           num_masks: int,
+                           shape: Tuple,
+                           dropout_rate: float):
+    if dropout_rate==0:
```
Comment (marcvanzee) [inline examples/cfq/models.py]:
```suggestion
    if dropout_rate == 0:
```

File: examples/cfq/models.py [hunk 9]
```diff
@@ -91,18 +100,99 @@ def apply(self,
...
+      return [None] * num_masks
+    masks = []
+    for i in range(0, num_masks):
+      # should I use a different rng for each mask?
```
Comment (marcvanzee) [inline examples/cfq/models.py]: Hmm not sure but I think so, otherwise all the masks could be generated from the same generator. But it would be good to ask this on FLAXers.

File: examples/cfq/models.py [hunk 10]
```diff
@@ -91,18 +100,99 @@ def apply(self,
...
+      dropout_mask = random.bernoulli(nn.make_rng(),
+                                      p=1 - dropout_rate, shape=shape)
+      # Scale mask.
```
Comment (marcvanzee) [inline examples/cfq/models.py]: Did you already figure out why we have to scale the mask?

File: examples/cfq/models.py [hunk 11]
```diff
@@ -91,18 +100,99 @@ def apply(self,
...
+      states.append(state)
+      cell_input = output
+      final_output = output
+    return states, final_output
+
+
 class Decoder(nn.Module):
   """LSTM decoder."""
 
+  def create_dropout_masks(self,
```
Comment (marcvanzee) [inline examples/cfq/models.py]: I think this function does not use `self`, so you could factor it out into a separate function that does not reside in this class.

File: examples/cfq/models.py [hunk 12]
```diff
@@ -111,24 +201,41 @@ def apply(self,
...
       carry_rng, categorical_rng = jax.random.split(rng, 2)
-      if not teacher_force:
+      if not train:
         x = last_prediction
       x = shared_embedding(x)
       dec_prev_state = jnp.expand_dims(h, 1)
       attention = mlp_attention(dec_prev_state, projected_keys,
                                 encoder_hidden_states, attention_mask)
-      lstm_state, y = lstm_cell((c, attention), x)
+      previous_states[0] = (previous_states[0][0],attention)
```
Comment (marcvanzee) [inline examples/cfq/models.py]: Please add a comment here explaining why you do this.

File: examples/cfq/models.py [hunk 13]
```diff
@@ -111,24 +201,41 @@ def apply(self,
...
+      
+      states, y = multilayer_lstm_cell(horizontal_dropout_masks=h_dropout_masks,
+                                       vertical_dropout_masks=v_dropout_masks,
+                                       input=x,
+                                       previous_states=previous_states)
       logits = projection(y)
       predicted_tokens = jax.random.categorical(categorical_rng, logits)
       predicted_tokens_uint8 = jnp.asarray(predicted_tokens, dtype=jnp.uint8)
-      return (carry_rng, lstm_state,
+      return (carry_rng, (states, y),
```
Comment (marcvanzee) [inline examples/cfq/models.py]: Why do you pass `(states, y)`, while in the previous version you only passed `state`?

File: examples/cfq/models.py [hunk 14]
```diff
@@ -111,24 +201,41 @@ def apply(self,
...
               predicted_tokens_uint8), (logits, predicted_tokens_uint8)
 
-    init_carry = (nn.make_rng(), init_state, inputs[:, 0])
+    init_states = [init_state] * num_layers
+    multilayer_lstm_output = (init_states, init_state[1])
```
Comment (marcvanzee) [inline examples/cfq/models.py:239]: Why do you use `(init_states, init_state[1])`? Could you add a comment?

File: examples/cfq/models.py [hunk 15]
```diff
@@ -147,13 +254,18 @@ def compute_attention_masks(mask_shape: Tuple, lengths: jnp.array):
...
   return mask
 
-
 class Seq2seq(nn.Module):
   """Sequence-to-sequence class using encoder/decoder architecture."""
 
-  def _create_modules(self, hidden_size):
-    encoder = Encoder.partial(hidden_size=hidden_size).shared(name='encoder')
-    decoder = Decoder.shared(name='decoder')
+  def _create_modules(self, hidden_size, num_layers, dropout):
```
Comment (marcvanzee) [inline examples/cfq/models.py]: This can be simplified, I merged this PR yesterday: https://github.com/google/flax/pull/468

File: flax/nn/recurrent.py [hunk 16]
```diff
@@ -15,6 +15,7 @@
 import jax
 import jax.numpy as jnp
 from flax import jax_utils
+from flax import nn
```
Comment (marcvanzee) [inline flax/nn/recurrent.py:18]: Why did you add this?

**Review by marcvanzee — CHANGES_REQUESTED** [review]
Looks good! I have a few comments, I think it would be useful to add some comments at various places to make the code easier to follow.

File: examples/cfq/models.py (hunk 13 above)
Comment (Ranapop) [reply examples/cfq/models.py]: I need to pass the state at each layer, as I am keeping track of them outside of the MultilayerLSTM module. Also, I need to pass around y as well because since I have dropout the hidden state and the output now differ

File: examples/cfq/models.py (hunk 10 above)
Comment (Ranapop) [reply examples/cfq/models.py]: not yet

File: examples/cfq/models.py (hunk 11 above)
Comment (Ranapop) [reply examples/cfq/models.py]: since it's only used in Decoder, I would not move it out entirely, but can annotate it as static

File: examples/cfq/models.py (hunk 4 above)
Comment (Ranapop) [reply examples/cfq/models.py]: It should still be shared across time steps (MultilayerLSTM will be called multiple times in the Decoder's apply function, hence this LSTMCell as well, I am not sure whether it's enough to have the MultilayerLSTM as shared)

File: examples/cfq/models.py (hunk 10 above)
Comment (Ranapop) [reply examples/cfq/models.py]: Before the scaling the mask is just an array of boolean values - after the scaling it becomes a probability distribution of floats

**Review by bohnetbd — APPROVED** [review]

File: examples/cfq/models.py [hunk 17]
```diff
@@ -91,22 +97,102 @@ def apply(self,
...
+        layer (used only between layers, not on final output). The dropout is
+        the one used here https://arxiv.org/abs/1512.05287 except it's not
+        applied on the first layer inputs and final outputs to be consistent
+        with the implementation in `recurrent.py`
+        shape [num_layers-1, batch_size, hidden_size]
+      input: input given to the first LSTM layer [batch_size, input_size]
+      previous_states: list of (c,h) for each layer
+        shape [num_layers, batch_size, 2*hidden_size]
+    """
+    input
```
Comment (marcvanzee) [inline examples/cfq/models.py]:
```suggestion
```

File: examples/cfq/models.py (hunk 4 above)
Comment (marcvanzee) [reply examples/cfq/models.py]: If you want to share the module then it should have the same name. Now you are creating a new "shared" module with a different name, so effectively no sharing is happening. There is an example of this behavior in the last cell of the Colab: https://colab.sandbox.google.com/drive/1ZoxgJjJpPmsbYs14fsdG3EPm4gJhzy0r#scrollTo=F_yB1jUyQ_oe

File: examples/cfq/models.py (hunk 9 above)
Comment (marcvanzee) [reply examples/cfq/models.py]: Did you ask it on FLAXers? What did they reply?

File: examples/cfq/models.py (hunk 10 above)
Comment (marcvanzee) [reply examples/cfq/models.py]: Thanks for the explanation. could you please add the explanation in a comment here?

**Review by marcvanzee — CHANGES_REQUESTED** [review]

File: examples/cfq/models.py (hunk 4 above)
Comment (Ranapop) [reply examples/cfq/models.py]: It has the same name at the same layer which is what I want - I want sharing at the same layer but no sharing between layers

File: examples/cfq/models.py (hunk 9 above)
Comment (Ranapop) [reply examples/cfq/models.py]: I thought it might not be worth asking - I am thinking if I am using the same key I will be getting the same random distribution, which is not what I want. I guess I can put it in a colab and test it out but it feels like a safe assumption

**Review by marcvanzee — APPROVED** [review]
